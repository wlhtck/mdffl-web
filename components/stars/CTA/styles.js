import { mergeAll } from 'lodash/fp';
import { primary, white } from '../../nebulas/colors';
import fontStyles from '../../nebulas/Typography/styles';

const backgroundColors = { primary, secondary: white };
const fontColors = { primary: 'white', secondary: 'primary' };

const buttonStyles = ({ type = 'primary' } = {}) => mergeAll([
  fontStyles({
    family: 'raleway',
    size: '16px',
    weight: 800,
    color: fontColors[type],
    lineHeight: 1,
  }), {
    backgroundColor: backgroundColors[type],
    display: 'inline-block',
    padding: '16px',
    border: 'none',
    borderRadius: '4px',
    cursor: 'pointer',
    textDecoration: 'none',
  },
]);

export default buttonStyles;
