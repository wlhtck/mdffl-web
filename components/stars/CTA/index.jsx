import {
  string, func, bool, oneOf,
} from 'prop-types';
import { omit } from 'lodash/fp';
import {
  compose, mapProps, componentFromProp, branch, withProps, setDisplayName,
} from 'recompose';
import withStyles from '../../util/withStyles';
import styles from './styles';

const omitProps = compose(mapProps, omit);
const withExternalLinkProps = withProps({ rel: 'noreferrer noopener', target: '_blank' });
const asCTA = compose(omitProps(['href']), withProps(() => ({ component: 'button' })));
const asLink = compose(
  omitProps(['onClick']),
  withProps(() => ({ component: 'a' })),
  branch(
    ({ external }) => !!external,
    withExternalLinkProps,
  ),
);

const CTA = compose(
  withStyles(styles),
  branch(
    ({ href }) => !href,
    asCTA,
    asLink,
  ),
  omitProps(['external']),
  setDisplayName('CTA'),
)(componentFromProp('component'));

CTA.propTypes = {
  children: string.isRequired,
  className: string,
  external: bool,
  onClick: func,
  type: oneOf(['primary', 'secondary']),
  url: string,
};

CTA.defaultProps = {
  className: '',
  external: false,
  onClick: null,
  type: 'primary',
  url: null,
};

export default CTA;
