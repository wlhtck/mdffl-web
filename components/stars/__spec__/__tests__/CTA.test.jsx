import React from 'react';
import { render } from 'enzyme';
import CTA from '../../CTA';

describe('components/starts/CTA', () => {
  it('renders an interal link', () => {
    expect(render(<CTA href="/about">CTA</CTA>)).toMatchSnapshot();
  });

  it('renders an external link', () => {
    expect(render(<CTA href="https://google.com" external>CTA</CTA>)).toMatchSnapshot();
  });

  it('renders a button', () => {
    expect(render(<CTA onClick={() => {}}>CTA</CTA>)).toMatchSnapshot();
  });

  it('renders a secondary button', () => {
    expect(render(<CTA type="secondary">CTA</CTA>)).toMatchSnapshot();
  });
});
