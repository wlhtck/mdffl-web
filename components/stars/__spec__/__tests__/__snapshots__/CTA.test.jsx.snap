// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/starts/CTA renders a button 1`] = `
<button
  class="withStyles-u5urts-0 gqQEcA"
  type="primary"
>
  CTA
</button>
`;

exports[`components/starts/CTA renders a secondary button 1`] = `
<button
  class="withStyles-u5urts-0 hpEXTr"
  type="secondary"
>
  CTA
</button>
`;

exports[`components/starts/CTA renders an external link 1`] = `
<a
  class="withStyles-u5urts-0 gqQEcA"
  href="https://google.com"
  rel="noreferrer noopener"
  target="_blank"
  type="primary"
>
  CTA
</a>
`;

exports[`components/starts/CTA renders an interal link 1`] = `
<a
  class="withStyles-u5urts-0 gqQEcA"
  href="/about"
  type="primary"
>
  CTA
</a>
`;
